cmd/statusd/sync.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/status-im/status-go/geth/common"
+)
+
+func createContextFromTimeout(timeout int) (context.Context, context.CancelFunc) {
+	if timeout == 0 {
+		return context.WithCancel(context.Background())
+	}
+
+	return context.WithTimeout(context.Background(), time.Duration(timeout)*time.Minute)
+}
+
+// syncAndStopNode tries to sync the blockchain and stop the node.
+// It returns an exit code (`0` if successful or `1` in case of error)
+// that can be used in `os.Exit` to exit immediately when the function returns.
+// The special exit code `-1` is used if execution was interrupted.
+func syncAndStopNode(interruptCh <-chan struct{}, nodeManager common.NodeManager, timeout int) (exitCode int) {
+	log.Printf("syncAndStopNode: node will synchronize the chain and exit (timeout %d mins)", timeout)
+
+	ctx, cancel := createContextFromTimeout(timeout)
+	defer cancel()
+
+	doneSync := make(chan struct{})
+	errSync := make(chan error)
+	go func() {
+		if err := nodeManager.EnsureSync(ctx); err != nil {
+			errSync <- err
+		}
+		close(doneSync)
+	}()
+
+	select {
+	case err := <-errSync:
```

Comment:
Where is this closed? I think this might be blocking.

Reply:
It is never closed and that is fine. Go will garbage collect it when not used. It may never return but as it is in `select` statement and `doneSync` is always finally closed, it won't result in a deadlock. 

The whole code may block only if the timeout is set to `0` and the blockchain sync never finishes.

Reply:
> The whole code may block only if the timeout is set to 0 and the blockchain sync never finishes.

But in such a case, one can hit `ctrl+c` and `interruptCh` will return.

Reply:
You're right, I didn't know this would work: https://play.golang.org/p/URUZ3YdOeRL

Reply:
Yeah, it is super useful. Also, the fact that you [don't need to necessarily close a channel](https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open) is not obvious. And one more interesting and sometimes useful thing related to `select` is [default selection](https://tour.golang.org/concurrency/6).

Reply:
Thanks! `default` case is always very helpful indeed.
And even though I knew it wouldn't leak anything after termination, I wasn't aware that it would be garbage collected during runtime. That's a really nice detail to know. :)

The reason I was suspicious here is because, so far, I have mostly used `for` and `select` together and break `Loop:` from `case <-done:`.